Added in Support for Grabbing Current Operations

garrett-griffin opened the pull request:
- Enhanced the Operations and Flow objects
- Added a BusinessOperation object to represent the return from the /operations api call.
- Added a few entries to the Recipe enum (snekkja operations, cog operations, etc.) that were present on the /operations page

grgisme committed:
Updated the version in the poetry file, so that projects that use this can specify the new version as the minimum requirements, so that items like "Item" from pymerc.api.models.common are available (they are not available in 0.3.0).

grgisme committed:
Merge branch 'master' of https://github.com/garrett-griffin/pymerc

# Conflicts:
#	pyproject.toml

grgisme committed:
Updating Businesses model to allow for account_ids that are strings and not integers, as my account_id is alphanumeric.

grgisme committed:
Updating Businesses model to allow for account_ids that are strings and not integers, as my account_id is alphanumeric.

Also updating contract_ids to be optional, as if you don't have contracts yet, or have never accepted a contract, it doesn't seem to be present in the API.

grgisme committed:
Updating sponsor_id to be a string, as my account id is alphanumeric.

grgisme committed:
Fixed some issues with building.py and added in support for calculating the labor based on shortages in inputs.

TODO:
- Add in ships and carts to the calculation
- Factor in produced labor and HoH and Apprentices, do so into their actual buildings in case some labor is being wasted that way.
- Factor in cases where lack of storage space is preventing full operation.

grgisme committed:
Merging Josh's changes in.

grgisme committed:
Undid some of the work in static.py as the values for "slaked lime" are now consistent.

grgisme committed:
Adding in support for grabbing building operations.

grgisme committed:
Merge remote-tracking branch 'jmgilman/master'

# Conflicts:
#	pymerc/game/player.py

grgisme committed:
Polishing Operations

jmgilman commented on pymerc/game/recipe.py:
```diff
@@ -0,0 +1,93 @@
+
+    async def load(self):
+        """Loads the data for the recipe and calculates labor required."""
+        if self.data is None:
```
I would omit this. The idea is that someone can call `load()` repeatedly if they need to obtain up-to-date information.

jmgilman commented on pymerc/game/recipe.py:
```diff
@@ -0,0 +1,93 @@
+        else:
+            self.labor = self._determine_labor_required()
+
+    def _determine_labor_required(self) -> float:
```
I was starting to add some `@property`'s  to the models. We could add a `labour` property to the `Recipe` model that perform this logic.

jmgilman commented on pymerc/game/recipe.py:
```diff
@@ -0,0 +1,93 @@
+                return input_ingredient.amount
+        return 0.0
+
+    def get_labor(self) -> Optional[float]:
```
This is unnecessary. Labor can be accessed at `Recipe.labour`.

jmgilman commented on pymerc/game/recipe.py:
```diff
@@ -0,0 +1,93 @@
+        Returns:
+            float: The labor required for the target multiplier.
+        """
+        inventory_assets = inventory_assets or {}
```
You can just set the default value to `{}` in the function arguments. Same for the line below this one.

jmgilman commented on pymerc/game/recipe.py:
```diff
@@ -0,0 +1,93 @@
+        Returns:
+            float: The labor required for the target multiplier.
+        """
+        inventory_assets = list(building.items.values()) if building.items else []
```
Should we just throw an error if the building doesn't have managers/assets? Not sure I understand the behavior if they are absent.

jmgilman commented on pymerc/game/recipe_manager.py:
```diff
@@ -0,0 +1,39 @@
+from pymerc.game.recipe import Recipe
+
+
+class RecipeManager:
```
I think this class might be unnecessary. My concern is that using it actually ends up calling `get_recipes()` on the static API for each recipe, which is a very expensive call. Ideally, we should only load once, if possible. 

A way I was already thinking of solving this is by using caching. Adding [this decorator](https://docs.python.org/dev/library/functools.html#functools.lru_cache) to the static API functions will cache the results after the first call to them. Since the data really shouldn't change, I think this is safe.

grgisme committed:
- Added async_lru dependency.
- Cached the API calls of the static elements
- Ditched recipe_manager
- Tightened up the code in recipe.py and building.py
